core/src/test/java/org/apache/accumulo/core/zookeeper/ZooCacheTest.java

```diff
@@ -16,86 +16,118 @@
+      watcher.process(event);
+    }
+
+    @Override
+    long getZKClientObjectVersion() {
+      return 1L;
+    }
+
+  }
+
+  private static final String root =
+      Constants.ZROOT + "/" + UUID.randomUUID().toString() + Constants.ZTSERVERS;
+  private static final String ZPATH = root + "/testPath";
   private static final byte[] DATA = {(byte) 1, (byte) 2, (byte) 3, (byte) 4};
   private static final List<String> CHILDREN = java.util.Arrays.asList("huey", "dewey", "louie");
 
   private ZooSession zk;
-  private ZooCache zc;
+  private TestZooCache zc;
 
   @BeforeEach
   public void setUp() {
     zk = createStrictMock(ZooSession.class);
-    zc = new ZooCache(zk);
+    zc = new TestZooCache(zk, Set.of(root));
   }
 
+  @SuppressWarnings("unchecked")
   @Test
-  public void testGet() throws Exception {
-    testGet(false);
+  public void testOverlappingPaths() throws Exception {
+    expect(zk.getConnectionCounter()).andReturn(2L).times(2);
+    zk.addPersistentRecursiveWatchers(isA(Set.class), isA(Watcher.class));
+    replay(zk);
+    assertThrows(IllegalArgumentException.class,
+        () -> new ZooCache(zk, Set.of(root, root + "/localhost:9995")));
+
+    Set<String> goodPaths = Set.of("/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/compactors",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/dead/tservers",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/gc/lock",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/managers/lock",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/namespaces",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/recovery",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/root_tablet",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/sservers",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/tables",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/tservers",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/users",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/mini",
+        "/accumulo/8247eee6-a176-4e19-baf7-e3da965fe050/monitor/lock");
+    new ZooCache(zk, goodPaths);
+    verify(zk);
   }
 
   @Test
-  public void testGet_FillStat() throws Exception {
-    testGet(true);
-  }
-
-  private void testGet(boolean fillStat) throws Exception {
-    ZcStat myStat = null;
-    if (fillStat) {
-      myStat = new ZcStat();
-    }
-    final long ephemeralOwner = 123456789L;
-    Stat existsStat = new Stat();
-    existsStat.setEphemeralOwner(ephemeralOwner);
-    expect(zk.exists(eq(ZPATH), anyObject(Watcher.class))).andReturn(existsStat);
-    expect(zk.getData(eq(ZPATH), anyObject(Watcher.class), eq(existsStat))).andReturn(DATA);
+  public void testGet() throws Exception {
+    expect(zk.getData(ZPATH, null, new Stat())).andReturn(DATA);
     replay(zk);
 
     assertFalse(zc.dataCached(ZPATH));
-    assertArrayEquals(DATA, (fillStat ? zc.get(ZPATH, myStat) : zc.get(ZPATH)));
+    assertSame(DATA, zc.get(ZPATH));
```

Comment:
A lot of the code that was removed related to zk.exists calls seems fine.  One thing that was removed was testing that ZK stats data is cached when getting data.  Played around w/ adding this testing back and came up with the following.

```java
  @Test
  public void testGet() throws Exception {

    expect(zk.getData(eq(ZPATH), isNull(), isA(Stat.class))).andAnswer(()->{
      ((Stat)EasyMock.getCurrentArguments()[2]).setEphemeralOwner(123456789);
      return DATA;
    });
    replay(zk);

    assertFalse(zc.dataCached(ZPATH));
    ZcStat zcStat = new ZcStat();
    assertSame(DATA, zc.get(ZPATH, zcStat));
    assertEquals(123456789, zcStat.getEphemeralOwner());
    assertSame(DATA, zc.get(ZPATH));
    verify(zk);
  }
```

Reply:
Added test in ddeee6b